### ISSUE TITLE
More information with timeout

### ISSUE THREAD

**Issue by davidbrochart** [asphalt-framework/asphalt#48]
Currently, when a component requests a resource that never comes, the application times out without any information as to which component failed to get the resource:
```
ERROR:asphalt.core.runner:Timeout waiting for the root component to start
```
I'm not sure if it's currently possible to give more details. If not, it would be a great enhancement.

**Comment by agronholm** [comment]
This is among the greatest woes on Asphalt and is scheduled to be rectified in v5.0. Unfortunately, component initialization is precisely the part I have not yet figured out among the enhancements I've made to the 5.0 series. I've had scarcely any time lately to devote to this project.

**Comment by agronholm** [comment]
I will of course release it when I've figured out how component initialization should work. Here are some of the issues I'm struggling with:

- Component initialization does not work with task groups, as setup and tear-down happen in different tasks
- Container components (unlike regular components) cannot be data classes
- There is no way for components to wait for other components to supply them information before they start (examples: py4j jar paths, web server routes)
- Resources published by child components should only be visible to their parent and sibling components, not globally to all components, unless the parent component explicitly exports them. This would enable complex components to be easily grafted into an app without having to worry about namespace conflicts.
- Using two components of the same type is currently a bit clumsy; I want to have something like `sqlalchemy/master` and `sqlalchemy/read` in the config which would both use the `sqlalchemy` entry point to find the component

**Comment by agronholm** [comment]
> > Component initialization does not work with task groups, as setup and tear-down happen in different tasks
> 
> So does it make sense to use AnyIO?

v5.0 is fully based on AnyIO and structured concurrency. I just need to deal with the component design issues before I can get that release out the door.

**Comment by agronholm** [comment]
If you have any ideas regarding the third item (`There is no way for components to wait for other components to supply them information before they start (examples: py4j jar paths, web server routes)`), I'm willing to accept ideas. I've had several of my own but I shot them all down already.

**Comment by agronholm** [comment]
Suppose you have a container component, and two child components in it: the `py4j` component and a custom component that needs to run Java code. The custom component needs to add a path to its `.jar` file in the `py4j` service in order for it to work. The py4j component cannot add jar paths after the service has been initialized. How would this work?

**Comment by davidbrochart** [comment]
If several components must supply information to another component then it seems impossible for this component to know when it should start, or wait for information from other components, since it doesn't know which component is the last one. Or there should be a step before starting components, where a component (A) could register itself into another component (B) as an information provider. Then B would call all its information providers before starting.
But if only one component (A) should supply information to another component (B) before B starts, then it seems to me that A depends on B, and instead of A requesting the resource added by B, A should request a factory for creating this resource. Then A can create the resource with the needed information.

**Comment by agronholm** [comment]
I've been working on refactoring component initialization and I came to the realization that trying to redesign the component system around that requirement is probably a lost cause. I'm hacking it today and tomorrow and if the outcome is at least halfway sensible, I'll commit it.

**Comment by agronholm** [comment]
I've been agonizing over the component design for these past weeks and haven't figured out a solution, so I'll try rubber ducking the problems here.

The problems with the current component system are:

1. This issue (if root component startup times out, it won't tell which component(s) were left waiting for a resource)
1. Container components cannot be data classes because the superclass has an `__init__()` method
1. Container component's child components should not be able to add resources to the root context (this breaks the use case of providing general use container components)
1. `Component.start()` does not support the use of AnyIO/Trio flavor of task groups for running background service tasks
1. When overriding `ContainerComponent.start()`, it's far too easy to miss the superclass method call

I think I can figure out the first one, but the rest are hard nuts to crack.

To detect startup tasks left waiting for resources, I can get a list of tasks and their frames, and compile a list of which task is waiting on what. That should make it considerably easier to debug timeouts.

To make it possible for container components to be data classes, I have to scrap the `__init__()` method. But what would `add_component()` do then?

I can make an isolated context for initializing a container component's child components, but what context would I use for the container component's `start()` method? If I use the same context as for the child components, it won't be able to publish any resources for its sibling components. Likewise, if I use the context used for its siblings, it won't be able to use the resources published by its child components. I may have to rethink the context system entirely.

I wanted to merge `ContainerComponent` into `Component`, but then `Component.start()` would be obligated to always make the superclass call, which currently it doesn't need to do. To make this work, the component initialization code would have to be moved elsewhere.

**Comment by davidbrochart** [comment]
4. Should the root component provide a "global" task group, accessible from all contexts, that a component could use to run background tasks?

**Comment by davidbrochart** [comment]
> * There is no way for components to wait for other components to supply them information before they start (examples: py4j jar paths, web server routes)

I think there should be an optional extra-step when starting components. Currently, a component runs code when starting and optionally when stopping (via tear-down). If there were an optional "init" step before "start", a component could add whatever resource is needed by a sibling component, which could wait for this "init" step to be finished (meaning that all sibling components completed the "init" step) before continuing to the "start" step.

**Comment by agronholm** [comment]
> 4. Should the root component provide a "global" task group, accessible from all contexts, that a component could use to run background tasks?

I'm not sure that this is a good idea. I toyed with the same idea some time ago, but I concluded that any background tasks started within a context of a component should also stop when that context is torn down. Otherwise the resources they were injected from that context would outlive their contexts of origin. Imagine a database connection pool that gets set up within a context of a component. Then, that components spawns a background task in the root task group, and the task requests resources from the context of the origin component. Then the origin component and its context are shut down, but the background task keeps going. See the problem?

**Comment by agronholm** [comment]
> I think there should be an optional extra-step when starting components. Currently, a component runs code when starting and optionally when stopping (via tear-down). If there were an optional "init" step before "start", a component could add whatever resource is needed by a sibling component, which could wait for this "init" step to be finished (meaning that all sibling components completed the "init" step) before continuing to the "start" step.

I was thinking along the same lines. But I'm not sure that I have a real-world use case that warrants this, so I think I'm going to shelve this idea.

**Comment by davidbrochart** [comment]
I have a use case where a component tries to [get a resource](https://github.com/jupyter-server/jupyverse/blob/9483f60951a2c08b80c48b8c7c9cee9b0e64d30e/plugins/lab/fps_lab/main.py#L19) that may or may not have been added. The resource may be added indeed, but maybe not before this call to `get_resource`, because it ultimately depends on the event loop scheduling, right? I could await `request_resource` instead, but when should I eventually stop requesting it and decide that it was not added?
A "setup" checkpoint would at least guarantee that beyond this point, this resource was added or not.

**Comment by agronholm** [comment]
Yeah, `request_resource()` is problematic in the sense that it might wait forever for a resource that won't ever be there. My latest idea regarding this function was to only have it wait if used in the component startup phase (which has a timeout anyway), and otherwise fail immediately if there's no resource available.

**Comment by davidbrochart** [comment]
That would work indeed.

**Comment by agronholm** [comment]
I have a few projects I urgently need to get major releases out on, but after that I think I can finally focus on getting Asphalt 5 to the finish line. It will be glorious!

**Comment by davidbrochart** [comment]
Awesome! I think it's a great project, and I'm looking forward to the rework around AnyIO.
Regarding point 4, it seems that what is needed is something like `ctx.start_soon()`, where it will use the context's task group? That would ensure that any background task launched by a component would be cancelled when the context is shut down.

**Comment by davidbrochart** [comment]
@agronholm do you think you could publish your WIP branch of v5.0 so that I can test it and maybe contribute, or is it not in a functional state? Just asking, I know you've been busy on other projects.
